### PR TITLE
[SPARK-18063][SQL] Failed to infer constraints over multiple aliases

### PR DESCRIPTION
**Pull request by jiangxb1987**
## What changes were proposed in this pull request?

The `UnaryNode.getAliasedConstraints` function fails to replace all expressions by their alias where constraints contains more than one expression to be replaced. 
For example:

```
val tr = LocalRelation('a.int, 'b.string, 'c.int)
val multiAlias = tr.where('a === 'c + 10).select('a.as('x), 'c.as('y))
multiAlias.analyze.constraints
```

currently outputs:

```
ExpressionSet(Seq(
    IsNotNull(resolveColumn(multiAlias.analyze, "x")),
    IsNotNull(resolveColumn(multiAlias.analyze, "y"))
)
```

The constraint `resolveColumn(multiAlias.analyze, "x") === resolveColumn(multiAlias.analyze, "y") + 10)` is missing.
## How was this patch tested?

Add new test cases in `ConstraintPropagationSuite`.

### CODE REVIEW

**Comment by rxin** [comment]
cc @sameeragarwal

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala [hunk 1]
```diff
@@ -293,15 +293,19 @@ abstract class UnaryNode extends LogicalPlan {
    * expressions with the corresponding alias
    */
   protected def getAliasedConstraints(projectList: Seq[NamedExpression]): Set[Expression] = {
-    projectList.flatMap {
+    var aliasedConstraints = child.constraints.toSet
```
Comment (sameeragarwal) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala]: Remove `toSet` (`ExpressionSet` should support all these operations)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala [hunk 2]
```diff
@@ -293,15 +293,19 @@ abstract class UnaryNode extends LogicalPlan {
...
-        }).union(Set(EqualNullSafe(e, a.toAttribute)))
-      case _ =>
-        Set.empty[Expression]
-    }.toSet
+        })
+        aliasedConstraints += EqualNullSafe(e, a.toAttribute)
+      case _ => // Don't change.
+    }
+
+    aliasedConstraints
```
Comment (sameeragarwal) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala]: This should really return `aliasedConstraints -- child.constraints`. Perhaps rename `aliasedConstraints` to `allConstraints` and then return `allConstraints -- child.constraints`?

**Review by sameeragarwal — CHANGES_REQUESTED** [review]
Looks good modulo a couple of comments. Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala (hunk 1 above)
Comment (jiangxb1987) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala]: I transformed `child.constraints` to `Set[Expression]` as we need to apply `++=` operator on it.

**Comment by jiangxb1987** [comment]
@sameeragarwal I've addressed your comments! Thank you!

**Comment by sameeragarwal** [comment]
LGTM, thanks!

**Comment by sameeragarwal** [comment]
Just to be extra safe, it'd be great to re-run the test suite on this once  https://github.com/apache/spark/pull/15319 is merged.

**Comment by hvanhovell** [comment]
retest this please

**Comment by rxin** [comment]
Merging this in master/branch-2.0.

**Pull request closed without merging**